rfcs/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md

```diff
@@ -0,0 +1,379 @@
+> devices. However, we cannot control the degree of blurring and the current
+> behaviour does not blur the image sufficiently to obscure its contents. This
+> could lead to thinking that we could blur in our own thumbnail endpoint, but
+> that would be more resource intensive than using BlurHash (I believe). While
+> BlurHash is technically complex to implement, I do think it is the best
+> approach for reasons I will describe below.
+
+#### CSS filter
+
+This is the simplest option and only requires adding a `filter: blur(Xrem)`
+style to sensitive results.
+
+This option has the following downsides:
+
+- It is a basic Gaussian blur and may not sufficiently obscure the image if a
+  low number is used or may create an unsettling blob effect if a higher number
+  is used
+- Gaussian blur may not be as aesthetically pleasing as the alternative
+- It requires the image to be downloaded to be blurred by the client
+- Requires the client to blur the image which may not be viable for low-spec
+  devices like certain Chromebooks (important for education settings) or mobile
+  devices
+
+It has one significant upside in that it is supremely simple to implement, does
+not require any new dependencies, and would only require about few lines of
+frontend code rather than lots of frontend and API code to support it.
+Additionally, the image would quickly load after being unblurred by the user
+because it would just require removing the CSS style rather than getting the
+actual image.
+
+#### BlurHash
+
+This is a significantly more complex option that requires both API and frontend
+changes to support. Rather than sending the image to the client, the thumbnail
+URL would send a hashed version of the image that represents the blur. A client
+side library decodes the hash and displays it in place of the thumbnail. It
+would require two new dependencies, a JavaScript BlurHash decoder to render the
+hash and an API library to produce the hash from the thumbnail.
+
+It has the following downsides:
+
+- Significantly more complicated than Gaussian blur
+- Requires conditionally calculating the hash on the API side in the thumbnail
+  endpoint
+  - This would be done by using a new query parameter on the thumbnail endpoint
+    to request the blurred hash
+- Can have a performance impact on the thumbnail endpoint
+- Introduces a delay between when the image is unblurred when it is displayed
+  due to needing to download the actual thumbnail
+- Could have different performance issues if decoding the hash is a heavy
+  operation for low-spec clients
+  - Looking at the code, I highly doubt this because it's a
+    [fast bit of maths](https://github.com/mad-gooze/fast-blurhash/blob/main/index.js)
+    that (due to the nature of BlurHash) ignores almost all the content of the
+    image rather than a Gaussian blur which needs to consider ever aspect of the
+    image in order to apply the blur
```

Comment:
I don't have any comparisons, but it does require each image render to a canvas element, and I don't know the performance implications of having 20 canvas nodes and some (performant, small) decoding JS executing vs. many evocations of a css filter.

I did find an alternative solution that generates images but _also_ then expects you to apply a css blur on top of the generated placeholders. I'm not advocating for this solution but I found its README comparison of different approaches an interesting, but limited, read:

https://github.com/transitive-bullshit/lqip-modern

Reply:
Interesting, I like that approach. The examples are helpful for understanding what the unblurred images end up being. They're almost entirely unintelligible, especially for complex images.

> I don't know the performance implications of having 20 canvas nodes and some (performant, small) decoding JS executing vs. many evocations of a css filter.

I could measure this. My personal phone is on the far lower end of "mid-spec" and my personal laptop runs a single-core ARM process that only _just_ manages to render web video.

Maybe we could add a request for an implementation plan up front that requests:

1. Test each solution's performance and figure out which one actually performs best on low-spec hardware
2. Make a recommendation for which one to follow

The lqip approach doesn't look like it would be more work than BlurHash. If it has a significant performance advantage for low-spec hardware over BlurHash, that would make it a better candidate, even though it doesn't necessarily have the same long-term usability as BlurHash for other parts of Openverse.